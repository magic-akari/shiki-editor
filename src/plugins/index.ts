import type { BundledLanguage, BundledTheme } from "shiki";
import type { ShikiCode, ShikiOptions } from "../core.js";

export type IDisposable = () => void;
export type { ShikiCode } from "../core.js";

export interface PluginOptions extends ShikiOptions {
	readonly language: "text" | BundledLanguage;
	readonly theme: "none" | BundledTheme;
}

export type EditorPlugin = {
	(editor: ShikiCode, options: PluginOptions): IDisposable;
};

export * from "./autoload.js";
export * from "./closing_pairs.js";
export * from "./tab.js";
