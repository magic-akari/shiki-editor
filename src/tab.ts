import { ceilTab, floorTab, visibleWidthFromLeft, visibleWidthLeadingSpace } from "./common.js";

export interface IndentConfig {
	tabSize: number;
	insertSpaces: boolean;
}

export interface State {
	/**
	 * The whole text content.
	 */
	value: string;
	/**
	 * The start of the selection.
	 */
	selectionStart: number;
	/**
	 * The end of the selection.
	 */
	selectionEnd: number;
	/**
	 * The direction of the selection.
	 */
	selectionDirection?: "forward" | "backward" | "none";
}

export interface PatchAction {
	value: string;
	start: number;
	end: number;
	mode?: SelectionMode;
}

export interface SelectAction {
	start: number;
	end: number;
	direction?: "forward" | "backward" | "none";
}

export interface Action {
	/**
	 * The patched text content.
	 */
	patch?: PatchAction;
	/**
	 * The new selection.
	 */
	select?: SelectAction;
}

export function indentText(input: State, config: IndentConfig): Action {
	if (
		input.selectionStart !== input.selectionEnd &&
		(bothEndsSelected(input.value, input.selectionStart, input.selectionEnd) ||
			input.value.slice(input.selectionStart, input.selectionEnd).includes("\n"))
	) {
		return blockIndentText(input, config);
	}

	return simpleIndentText(input, config);
}

function simpleIndentText(input: State, config: IndentConfig): Action {
	const { value, selectionStart, selectionEnd } = input;
	const { tabSize, insertSpaces } = config;

	if (!insertSpaces) {
		return {
			patch: {
				value: "\t",
				start: selectionStart,
				end: selectionEnd,
				mode: "end",
			},
		};
	}

	const [width_from_left] = visibleWidthFromLeft(value, selectionStart, tabSize);
	const indent = " ".repeat(ceilTab(width_from_left + 1, tabSize) - width_from_left);
	return {
		patch: {
			value: indent,
			start: selectionStart,
			end: selectionEnd,
			mode: "end",
		},
	};
}

function blockIndentText(input: State, config: IndentConfig): Action {
	const { tabSize, insertSpaces } = config;
	const { value, selectionStart, selectionEnd, selectionDirection } = input;

	const block_start = getLineStart(value, selectionStart);
	const block_end = getLineEnd(value, selectionEnd);

	const block = value.slice(block_start, block_end);

	const replacement = block.replaceAll(/^[ \t]*/gm, (leading, offset, str) => {
		if (str[offset] === "\n" || str[offset] === "\r" || offset === block_end) return leading;

		let [tab_width] = visibleWidthFromLeft(leading, leading.length, tabSize, 0);
		tab_width = ceilTab(tab_width + 1, tabSize);

		if (insertSpaces) {
			return " ".repeat(tab_width);
		}
		return "\t".repeat(tab_width / tabSize);
	});

	const patch = {
		value: replacement,
		start: block_start,
		end: block_end,
		mode: "end",
	} satisfies PatchAction;

	// restore selection
	// By default, the selection is anchored at both ends.
	const select = {
		start: selectionStart,
		end: selectionEnd + replacement.length - block.length,
		direction: selectionDirection,
	} satisfies SelectAction;

	if (selectionStart !== block_start) {
		const line_start = block_start;
		const cursor_offset = selectionStart - line_start;
		const line = value.slice(line_start, selectionStart);
		const [, old_leading_offset] = visibleWidthLeadingSpace(line, tabSize);
		const [, max_leading_offset] = visibleWidthLeadingSpace(replacement, tabSize);

		if (cursor_offset > old_leading_offset) {
			// |<TAB>|<TAB>|T|ext
			//               ^
			select.start += max_leading_offset - old_leading_offset;
		} else if (cursor_offset > max_leading_offset) {
			// |     |     |Text (old)
			//           ^
			// |<TAB>|<TAB>|<TAB>|Text (new)
			//                   ^
			select.start = line_start + max_leading_offset;
		}
	}

	if (selectionEnd < block_end) {
		const line_start = getLineStart(value, selectionEnd);
		const cursor_offset = selectionEnd - line_start;
		const line = value.slice(line_start, selectionEnd);
		const [, old_leading_offset] = visibleWidthLeadingSpace(line, tabSize);

		const new_bottom_line_offset = getLineStart(replacement, replacement.length);
		const new_bottom_line = replacement.slice(new_bottom_line_offset);
		const [, max_leading_offset] = visibleWidthLeadingSpace(new_bottom_line, tabSize);

		if (cursor_offset <= old_leading_offset) {
			select.end = block_start + new_bottom_line_offset + cursor_offset;

			if (cursor_offset > max_leading_offset) {
				select.end += max_leading_offset - old_leading_offset;
			}
		}
	}

	return {
		patch,
		select,
	} satisfies Action;
}

export function outdentText(input: State, config: IndentConfig): Action {
	const { tabSize, insertSpaces } = config;
	const { value, selectionStart, selectionEnd, selectionDirection } = input;

	const block_start = getLineStart(value, selectionStart);
	const block_end = getLineEnd(value, selectionEnd);

	const block = value.slice(block_start, block_end);

	const replacement = block.replaceAll(/^[ \t]*/gm, (leading) => {
		let [tab_width] = visibleWidthFromLeft(leading, leading.length, tabSize, 0);
		tab_width = floorTab(tab_width - 1, tabSize);

		if (tab_width <= 0) return "";

		if (insertSpaces) {
			return " ".repeat(tab_width);
		}
		return "\t".repeat(tab_width / tabSize);
	});
	if (replacement === block) return {};

	const patch = {
		value: replacement,
		start: block_start,
		end: block_end,
		mode: "end",
	} satisfies PatchAction;

	// restore selection
	// By default, the selection is anchored at both ends.
	const select = {
		start: selectionStart,
		end: selectionEnd + replacement.length - block.length,
		direction: selectionDirection,
	} satisfies SelectAction;

	if (selectionStart !== block_start) {
		const line_start = block_start;
		const cursor_offset = selectionStart - line_start;
		const line = value.slice(line_start, selectionStart);
		const [, old_leading_offset] = visibleWidthLeadingSpace(line, tabSize);
		const [, max_leading_offset] = visibleWidthLeadingSpace(replacement, tabSize);

		if (cursor_offset > old_leading_offset) {
			// |<TAB>|<TAB>|T|ext
			//               ^
			select.start += max_leading_offset - old_leading_offset;
		} else if (cursor_offset > max_leading_offset) {
			// |     |     |Text (old)
			//           ^
			// |<TAB>|<TAB>|<TAB>|Text (new)
			//                   ^
			select.start = line_start + max_leading_offset;
		}
	}

	if (selectionEnd < block_end) {
		const line_start = getLineStart(value, selectionEnd);
		const cursor_offset = selectionEnd - line_start;
		const line = value.slice(line_start, selectionEnd);
		const [, old_leading_offset] = visibleWidthLeadingSpace(line, tabSize);

		const new_bottom_line_offset = getLineStart(replacement, replacement.length);
		const new_bottom_line = replacement.slice(new_bottom_line_offset);
		const [, max_leading_offset] = visibleWidthLeadingSpace(new_bottom_line, tabSize);

		if (cursor_offset <= old_leading_offset) {
			select.end = block_start + new_bottom_line_offset + cursor_offset;

			if (cursor_offset > max_leading_offset) {
				select.end += max_leading_offset - old_leading_offset;
			}
		}
	}

	return {
		patch,
		select,
	} satisfies Action;
}

function enter(input: State, config: IndentConfig): Action {
	if (input.selectionStart !== input.selectionEnd) {
		return {};
	}
	const { value, selectionStart } = input;
	const line_start = getLineStart(value, selectionStart);
	if (line_start === selectionStart) {
		return {};
	}

	const line = value.slice(line_start, selectionStart);
	let [leading_space] = visibleWidthLeadingSpace(line, config.tabSize);
	leading_space = floorTab(leading_space, config.tabSize);
	let indet_space = leading_space;

	switch (value[selectionStart - 1]) {
		case "(":
		case "[":
		case "{": {
			indet_space += config.tabSize;
		}
	}

	let replacement = "\n" + " ".repeat(indet_space);
	if (config.insertSpaces) {
		replacement = "\n" + "\t".repeat(indet_space / config.tabSize);
	}

	let select: SelectAction | undefined;

	switch (value.slice(selectionStart - 1, selectionStart + 1)) {
		case "{}":
		case "[]":
		case "()": {
			select = {
				start: selectionStart + replacement.length,
				end: selectionStart + replacement.length,
				direction: "none",
			};

			if (config.insertSpaces) {
				replacement += "\n" + " ".repeat(leading_space);
			} else {
				replacement += "\n" + "\t".repeat(leading_space / config.tabSize);
			}
		}
	}

	return {
		patch: {
			value: replacement,
			start: selectionStart,
			end: selectionStart,
			mode: "end",
		},
		select,
	};
}

function bothEndsSelected(text: string, start: number, end: number): boolean {
	const is_start = start === 0 || text[start - 1] === "\n";
	const is_end = end === text.length || text[end] === "\n" || text[end] === "\r";
	return start !== end && is_start && is_end;
}

function getLineStart(text: string, index: number): number {
	while (index > 0 && text[index - 1] !== "\n") {
		index--;
	}
	return index;
}

/**
 * Get the end index of the line.
 * - all suffix "\n" and "\r" are ignored.
 * - text[end] === "\n" || text[end] === "\r" || end === text.length
 * - text.slice(start, end) should not include "\n" at the end.
 */
function getLineEnd(text: string, index: number): number {
	if (text[index - 1] === "\n") {
		index--;
	}

	while (index > 0 && (index === text.length || text[index] === "\n" || text[index] === "\r")) {
		index--;
	}

	while (index < text.length && text[index] !== "\n" && text[index] !== "\r") {
		index++;
	}

	return index;
}

export function hookTab(input: HTMLTextAreaElement, config: IndentConfig) {
	const onKeydown = (e: KeyboardEvent) => {
		console.log(e.key);
		switch (e.key) {
			case "Tab": {
				const action = e.shiftKey ? outdentText : indentText;
				const { patch, select } = action(e.target as HTMLTextAreaElement, config);
				if (patch || select) {
					e.preventDefault();
				}
				if (patch) {
					input.setRangeText(patch.value, patch.start, patch.end, patch.mode);
					input.dispatchEvent(new Event("input"));
					input.dispatchEvent(new Event("change"));
				}
				if (select) {
					input.setSelectionRange(select.start, select.end, select.direction);
					input.dispatchEvent(new Event("selectionchange"));
				}
				break;
			}

			case "Enter": {
				const { patch, select } = enter(e.target as HTMLTextAreaElement, config);
				if (patch || select) {
					e.preventDefault();
				}
				if (patch) {
					input.setRangeText(patch.value, patch.start, patch.end, patch.mode);
					input.dispatchEvent(new Event("input"));
					input.dispatchEvent(new Event("change"));
				}
				if (select) {
					input.setSelectionRange(select.start, select.end, select.direction);
					input.dispatchEvent(new Event("selectionchange"));
				}
				break;
			}

			case "Backspace": {
				break;
			}

			default:
				return;
		}
	};

	input.addEventListener("keydown", onKeydown);
	return () => {
		input.removeEventListener("keydown", onKeydown);
	};
}
