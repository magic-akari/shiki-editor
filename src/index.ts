import { getHighlighter, type BundledLanguage, type BundledTheme } from 'shiki'
import { hookIndent, hookOutdent } from "./tab.js";
import { hookBracket } from "./bracket.js";

interface EditorOptions {
	value?: string;
	language: BundledLanguage;
	/**
	 * Control the rendering of line numbers.
	 * Defaults to `on`.
	 */
	lineNumbers?: "on" | "off";
	/**
	 * Should the editor be read only. See also `domReadOnly`.
	 * Defaults to false.
	 */
	readOnly?: boolean;

	theme?: BundledTheme;
	/**
	 * The number of spaces a tab is equal to.
	 * This setting is overridden based on the file contents when `detectIndentation` is on.
	 * Defaults to 4.
	 */
	tabSize?: number;
	/**
	 * Insert spaces when pressing `Tab`.
	 * This setting is overridden based on the file contents when `detectIndentation` is on.
	 * Defaults to true.
	 */
	insertSpaces?: boolean;
}

const defaultOptions: EditorOptions = {
	value: "",
	language: "c",
	lineNumbers: "on",
	readOnly: false,
	theme: "github-light",
	tabSize: 4,
	insertSpaces: true,
};

interface ICodeEditor {
	updateOptions(options: EditorOptions): void;
	/**
	 * Dispose the editor.
	 */
	dispose(): void;
}

export function create(domElement: HTMLElement, options?: EditorOptions): ICodeEditor {
	const doc = domElement.ownerDocument;

	const output = doc.createElement("div");
	const input = doc.createElement("textarea");

	output.classList.add("shiki-editor", "output");
	input.classList.add("shiki-editor", "input");

	domElement.appendChild(output);
	domElement.appendChild(input);

	const highlighter = getHighlighter({
		themes: ["github-light"],
		langs: ["javascript"],
	});

	const render = async () => {
		const { codeToTokens } = await highlighter;

		const codeToHtml = (code: string) => {
			const {
				tokens: tokensLines,
				fg, bg,
				themeName,
				rootStyle
			} = codeToTokens(code, {
				lang: "javascript",
				theme: "github-light",
			});
			const lines = tokensLines.map((tokenLine, index) => (`<span class="line">${
				tokenLine
					.map(token => `<span class="${
						`position:${index + 1}:${token.offset + 1},${token.offset + 1 + token.content.length}` + " " +
						`font-style:${token.fontStyle}`
					}" style="color: ${token.color}">${token.content}</span>`)
					.join("")
			}</span>`))
			return `<pre class="shiki ${themeName}" style="background-color:${bg};color:${fg};${rootStyle ? rootStyle : ''}" tabindex="0"><code>${lines}</code></pre>`
		}
		output.innerHTML = codeToHtml(input.value);
	};

	input.addEventListener("input", render);

	if (options?.value) {
		input.value = options.value;
	}
	render();

	const cleanup = [
		() => {
			input.removeEventListener("input", render);
		},
		hookIndent(input, options?.tabSize ?? 4, options?.insertSpaces ?? true),
		hookOutdent(input, options?.tabSize ?? 4, options?.insertSpaces ?? true),
		hookBracket(input),
		() => {
			domElement.removeChild(input);
			domElement.removeChild(output);
		},
	];

	return {
		updateOptions: (newOptions: EditorOptions) => {
			console.log("updateOptions", newOptions);
		},
		dispose() {
			cleanup.forEach((fn) => fn());
		},
	};
}
