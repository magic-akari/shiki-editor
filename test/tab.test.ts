import { Action, PatchAction, SelectAction, State, indentText, outdentText } from "../src/tab";

interface TabConfig {
	tabSize: number;
	insertSpaces: boolean;
}

const space_4 = {
	tabSize: 4,
	insertSpaces: true,
};

const tab_4 = {
	tabSize: 4,
	insertSpaces: false,
};

const space_2 = {
	tabSize: 2,
	insertSpaces: true,
};

const tab_2 = {
	tabSize: 2,
	insertSpaces: false,
};

function setRangeText(state: State, patch: PatchAction): State {
	const [start, mid, end] = [state.value.slice(0, patch.start), patch.value, state.value.slice(patch.end)];

	const new_state = { ...state };

	switch (patch.mode) {
		case "start":
			new_state.selectionStart = new_state.selectionEnd = start.length;
			break;
		case "end":
			new_state.selectionStart = new_state.selectionEnd = start.length + mid.length;
			break;
		case "select":
		default:
			// Unfortunately, we cannot mimic the behavior of `preserve`.
			new_state.selectionStart = start.length;
			new_state.selectionEnd = start.length + mid.length;
			break;
	}

	new_state.value = start + mid + end;
	return new_state;
}

function setSelectionRange(state: State, select: SelectAction): State {
	return { value: state.value, selectionStart: select.start, selectionEnd: select.end };
}

function applyEdit(state: State, action?: Action): State {
	if (!action) {
		return state;
	}
	if (action.patch) {
		state = setRangeText(state, action.patch);
	}
	if (action.select) {
		state = setSelectionRange(state, action.select);
	}
	return state;
}

const simpleText = "My First Line\n\t\tMy Second Line\n    Third Line\n\n123";

describe("Editor Commands - ShiftCommand", () => {
	test("basic usage", () => {
		testShiftCommand(simpleText, 0, 0, space_4);
	});

	test("basic usage", () => {
		testShiftCommand(simpleText, 2, 2, space_4);
	});

	test("basic usage", () => {
		const select = simpleText.indexOf("Third Line") - 2;
		testShiftCommand(simpleText, select, select, tab_2);
	});

	test("single line", () => {
		const select = "My First Line";
		testShiftCommand(simpleText, 0, select.length, space_4);
	});

	test("single line", () => {
		const select = "My First Line\n";
		testShiftCommand(simpleText, 0, select.length, space_4);
	});

	test("single line", () => {
		const select = "\t\tMy Second Line\n";
		const start = simpleText.indexOf(select);
		testShiftCommand(simpleText, start, start + select.length, space_4);
	});

	test("multiple lines", () => {
		const select = "First Line\n";
		const start = simpleText.indexOf(select);
		testShiftCommand(simpleText, start, start + select.length, space_4);
	});

	test("multiple lines", () => {
		const select = "Line\n\t\tMy";
		const start = simpleText.indexOf(select);
		testShiftCommand(simpleText, start, start + select.length, space_4);
	});

	test("multiple lines", () => {
		const select = "Line\n\t\tMy Second Line\n";
		const start = simpleText.indexOf(select);
		testShiftCommand(simpleText, start, start + select.length, space_4);
	});

	test("multiple lines", () => {
		const select = "First Line\n\t";
		const start = simpleText.indexOf(select);
		testShiftCommand(simpleText, start, start + select.length, space_4);
	});

	test("multiple lines", () => {
		const select = "\tMy Second Line\n  ";
		const start = simpleText.indexOf(select);
		testShiftCommand(simpleText, start, start + select.length, space_4);
	});

	test("multiple lines", () => {
		testShiftCommand(simpleText, 0, simpleText.length, space_4);
	});
});

describe("Editor Commands - UnShiftCommand", () => {
	test("basic usage", () => {
		testUnShiftCommand(simpleText, 0, 0, space_4);
	});

	test("basic usage", () => {
		testUnShiftCommand(simpleText, 2, 2, space_4);
	});

	test("basic usage", () => {
		const select = simpleText.indexOf("Third Line") - 2;
		testUnShiftCommand(simpleText, select, select, tab_2);
	});

	test("single line", () => {
		const select = "My First Line";
		testUnShiftCommand(simpleText, 0, select.length, space_4);
	});

	test("single line", () => {
		const select = "My First Line\n";
		testUnShiftCommand(simpleText, 0, select.length, space_4);
	});

	test("single line", () => {
		const select = "\t\tMy Second Line\n";
		const start = simpleText.indexOf(select);
		testUnShiftCommand(simpleText, start, start + select.length, space_4);
	});

	test("multiple lines", () => {
		const select = "First Line\n";
		const start = simpleText.indexOf(select);
		testUnShiftCommand(simpleText, start, start + select.length, space_4);
	});

	test("multiple lines", () => {
		const select = "Line\n\t\tMy";
		const start = simpleText.indexOf(select);
		testUnShiftCommand(simpleText, start, start + select.length, space_4);
	});

	test("multiple lines", () => {
		const select = "Line\n\t\tMy Second Line\n";
		const start = simpleText.indexOf(select);
		testUnShiftCommand(simpleText, start, start + select.length, space_4);
	});

	test("multiple lines", () => {
		const select = "First Line\n\t";
		const start = simpleText.indexOf(select);
		testUnShiftCommand(simpleText, start, start + select.length, space_4);
	});

	test("multiple lines", () => {
		const select = "\tMy Second Line\n  ";
		const start = simpleText.indexOf(select);
		testUnShiftCommand(simpleText, start, start + select.length, space_4);
	});

	test("multiple lines", () => {
		testUnShiftCommand(simpleText, 0, simpleText.length, space_4);
	});
});

function testShiftCommand(value: string, selectionStart: number, selectionEnd: number, config: TabConfig): void {
	const state = { value, selectionStart, selectionEnd };

	const action = indentText(state, config);
	const result = applyEdit(state, action);

	expect({ config, input: handleState(state), output: handleState(result) }).toMatchSnapshot();
}

function testUnShiftCommand(value: string, selectionStart: number, selectionEnd: number, config: TabConfig): void {
	const state = { value, selectionStart, selectionEnd };

	const action = outdentText(state, config);
	const result = applyEdit(state, action);

	expect({ config, input: handleState(state), output: handleState(result) }).toMatchSnapshot();
}

function handleState(state: State) {
	const lines = state.value.split(/\r\n|\n/);
	return {
		...state,
		value: lines,
	};
}
