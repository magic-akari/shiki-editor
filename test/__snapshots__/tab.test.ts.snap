// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`Editor Commands - ShiftCommand basic usage 1`] = `
{
  "config": {
    "insertSpaces": true,
    "tabSize": 4,
  },
  "input": {
    "selectionEnd": 0,
    "selectionStart": 0,
    "value": [
      "My First Line",
      "		My Second Line",
      "    Third Line",
      "",
      "123",
    ],
  },
  "output": {
    "selectionEnd": 4,
    "selectionStart": 4,
    "value": [
      "    My First Line",
      "		My Second Line",
      "    Third Line",
      "",
      "123",
    ],
  },
}
`;

exports[`Editor Commands - ShiftCommand basic usage 2`] = `
{
  "config": {
    "insertSpaces": true,
    "tabSize": 4,
  },
  "input": {
    "selectionEnd": 2,
    "selectionStart": 2,
    "value": [
      "My First Line",
      "		My Second Line",
      "    Third Line",
      "",
      "123",
    ],
  },
  "output": {
    "selectionEnd": 4,
    "selectionStart": 4,
    "value": [
      "My   First Line",
      "		My Second Line",
      "    Third Line",
      "",
      "123",
    ],
  },
}
`;

exports[`Editor Commands - ShiftCommand basic usage 3`] = `
{
  "config": {
    "insertSpaces": false,
    "tabSize": 2,
  },
  "input": {
    "selectionEnd": 33,
    "selectionStart": 33,
    "value": [
      "My First Line",
      "		My Second Line",
      "    Third Line",
      "",
      "123",
    ],
  },
  "output": {
    "selectionEnd": 34,
    "selectionStart": 34,
    "value": [
      "My First Line",
      "		My Second Line",
      "  	  Third Line",
      "",
      "123",
    ],
  },
}
`;

exports[`Editor Commands - ShiftCommand single line 1`] = `
{
  "config": {
    "insertSpaces": true,
    "tabSize": 4,
  },
  "input": {
    "selectionEnd": 13,
    "selectionStart": 0,
    "value": [
      "My First Line",
      "		My Second Line",
      "    Third Line",
      "",
      "123",
    ],
  },
  "output": {
    "selectionEnd": 17,
    "selectionStart": 0,
    "value": [
      "    My First Line",
      "		My Second Line",
      "    Third Line",
      "",
      "123",
    ],
  },
}
`;

exports[`Editor Commands - ShiftCommand single line 2`] = `
{
  "config": {
    "insertSpaces": true,
    "tabSize": 4,
  },
  "input": {
    "selectionEnd": 14,
    "selectionStart": 0,
    "value": [
      "My First Line",
      "		My Second Line",
      "    Third Line",
      "",
      "123",
    ],
  },
  "output": {
    "selectionEnd": 18,
    "selectionStart": 0,
    "value": [
      "    My First Line",
      "		My Second Line",
      "    Third Line",
      "",
      "123",
    ],
  },
}
`;

exports[`Editor Commands - ShiftCommand single line 3`] = `
{
  "config": {
    "insertSpaces": true,
    "tabSize": 4,
  },
  "input": {
    "selectionEnd": 31,
    "selectionStart": 14,
    "value": [
      "My First Line",
      "		My Second Line",
      "    Third Line",
      "",
      "123",
    ],
  },
  "output": {
    "selectionEnd": 41,
    "selectionStart": 14,
    "value": [
      "My First Line",
      "            My Second Line",
      "    Third Line",
      "",
      "123",
    ],
  },
}
`;

exports[`Editor Commands - ShiftCommand multiple lines 1`] = `
{
  "config": {
    "insertSpaces": true,
    "tabSize": 4,
  },
  "input": {
    "selectionEnd": 14,
    "selectionStart": 3,
    "value": [
      "My First Line",
      "		My Second Line",
      "    Third Line",
      "",
      "123",
    ],
  },
  "output": {
    "selectionEnd": 18,
    "selectionStart": 7,
    "value": [
      "    My First Line",
      "		My Second Line",
      "    Third Line",
      "",
      "123",
    ],
  },
}
`;

exports[`Editor Commands - ShiftCommand multiple lines 2`] = `
{
  "config": {
    "insertSpaces": true,
    "tabSize": 4,
  },
  "input": {
    "selectionEnd": 18,
    "selectionStart": 9,
    "value": [
      "My First Line",
      "		My Second Line",
      "    Third Line",
      "",
      "123",
    ],
  },
  "output": {
    "selectionEnd": 32,
    "selectionStart": 13,
    "value": [
      "    My First Line",
      "            My Second Line",
      "    Third Line",
      "",
      "123",
    ],
  },
}
`;

exports[`Editor Commands - ShiftCommand multiple lines 3`] = `
{
  "config": {
    "insertSpaces": true,
    "tabSize": 4,
  },
  "input": {
    "selectionEnd": 31,
    "selectionStart": 9,
    "value": [
      "My First Line",
      "		My Second Line",
      "    Third Line",
      "",
      "123",
    ],
  },
  "output": {
    "selectionEnd": 45,
    "selectionStart": 13,
    "value": [
      "    My First Line",
      "            My Second Line",
      "    Third Line",
      "",
      "123",
    ],
  },
}
`;

exports[`Editor Commands - ShiftCommand multiple lines 4`] = `
{
  "config": {
    "insertSpaces": true,
    "tabSize": 4,
  },
  "input": {
    "selectionEnd": 15,
    "selectionStart": 3,
    "value": [
      "My First Line",
      "		My Second Line",
      "    Third Line",
      "",
      "123",
    ],
  },
  "output": {
    "selectionEnd": 19,
    "selectionStart": 7,
    "value": [
      "    My First Line",
      "            My Second Line",
      "    Third Line",
      "",
      "123",
    ],
  },
}
`;

exports[`Editor Commands - ShiftCommand multiple lines 5`] = `
{
  "config": {
    "insertSpaces": true,
    "tabSize": 4,
  },
  "input": {
    "selectionEnd": 33,
    "selectionStart": 15,
    "value": [
      "My First Line",
      "		My Second Line",
      "    Third Line",
      "",
      "123",
    ],
  },
  "output": {
    "selectionEnd": 43,
    "selectionStart": 15,
    "value": [
      "My First Line",
      "            My Second Line",
      "        Third Line",
      "",
      "123",
    ],
  },
}
`;

exports[`Editor Commands - ShiftCommand multiple lines 6`] = `
{
  "config": {
    "insertSpaces": true,
    "tabSize": 4,
  },
  "input": {
    "selectionEnd": 50,
    "selectionStart": 0,
    "value": [
      "My First Line",
      "		My Second Line",
      "    Third Line",
      "",
      "123",
    ],
  },
  "output": {
    "selectionEnd": 72,
    "selectionStart": 0,
    "value": [
      "    My First Line",
      "            My Second Line",
      "        Third Line",
      "",
      "    123",
    ],
  },
}
`;

exports[`Editor Commands - UnShiftCommand basic usage 1`] = `
{
  "config": {
    "insertSpaces": true,
    "tabSize": 4,
  },
  "input": {
    "selectionEnd": 0,
    "selectionStart": 0,
    "value": [
      "My First Line",
      "		My Second Line",
      "    Third Line",
      "",
      "123",
    ],
  },
  "output": {
    "selectionEnd": 0,
    "selectionStart": 0,
    "value": [
      "My First Line",
      "		My Second Line",
      "    Third Line",
      "",
      "123",
    ],
  },
}
`;

exports[`Editor Commands - UnShiftCommand basic usage 2`] = `
{
  "config": {
    "insertSpaces": true,
    "tabSize": 4,
  },
  "input": {
    "selectionEnd": 2,
    "selectionStart": 2,
    "value": [
      "My First Line",
      "		My Second Line",
      "    Third Line",
      "",
      "123",
    ],
  },
  "output": {
    "selectionEnd": 2,
    "selectionStart": 2,
    "value": [
      "My First Line",
      "		My Second Line",
      "    Third Line",
      "",
      "123",
    ],
  },
}
`;

exports[`Editor Commands - UnShiftCommand basic usage 3`] = `
{
  "config": {
    "insertSpaces": false,
    "tabSize": 2,
  },
  "input": {
    "selectionEnd": 33,
    "selectionStart": 33,
    "value": [
      "My First Line",
      "		My Second Line",
      "    Third Line",
      "",
      "123",
    ],
  },
  "output": {
    "selectionEnd": 32,
    "selectionStart": 32,
    "value": [
      "My First Line",
      "		My Second Line",
      "	Third Line",
      "",
      "123",
    ],
  },
}
`;

exports[`Editor Commands - UnShiftCommand single line 1`] = `
{
  "config": {
    "insertSpaces": true,
    "tabSize": 4,
  },
  "input": {
    "selectionEnd": 13,
    "selectionStart": 0,
    "value": [
      "My First Line",
      "		My Second Line",
      "    Third Line",
      "",
      "123",
    ],
  },
  "output": {
    "selectionEnd": 13,
    "selectionStart": 0,
    "value": [
      "My First Line",
      "		My Second Line",
      "    Third Line",
      "",
      "123",
    ],
  },
}
`;

exports[`Editor Commands - UnShiftCommand single line 2`] = `
{
  "config": {
    "insertSpaces": true,
    "tabSize": 4,
  },
  "input": {
    "selectionEnd": 14,
    "selectionStart": 0,
    "value": [
      "My First Line",
      "		My Second Line",
      "    Third Line",
      "",
      "123",
    ],
  },
  "output": {
    "selectionEnd": 14,
    "selectionStart": 0,
    "value": [
      "My First Line",
      "		My Second Line",
      "    Third Line",
      "",
      "123",
    ],
  },
}
`;

exports[`Editor Commands - UnShiftCommand single line 3`] = `
{
  "config": {
    "insertSpaces": true,
    "tabSize": 4,
  },
  "input": {
    "selectionEnd": 31,
    "selectionStart": 14,
    "value": [
      "My First Line",
      "		My Second Line",
      "    Third Line",
      "",
      "123",
    ],
  },
  "output": {
    "selectionEnd": 33,
    "selectionStart": 14,
    "value": [
      "My First Line",
      "    My Second Line",
      "    Third Line",
      "",
      "123",
    ],
  },
}
`;

exports[`Editor Commands - UnShiftCommand multiple lines 1`] = `
{
  "config": {
    "insertSpaces": true,
    "tabSize": 4,
  },
  "input": {
    "selectionEnd": 14,
    "selectionStart": 3,
    "value": [
      "My First Line",
      "		My Second Line",
      "    Third Line",
      "",
      "123",
    ],
  },
  "output": {
    "selectionEnd": 14,
    "selectionStart": 3,
    "value": [
      "My First Line",
      "		My Second Line",
      "    Third Line",
      "",
      "123",
    ],
  },
}
`;

exports[`Editor Commands - UnShiftCommand multiple lines 2`] = `
{
  "config": {
    "insertSpaces": true,
    "tabSize": 4,
  },
  "input": {
    "selectionEnd": 18,
    "selectionStart": 9,
    "value": [
      "My First Line",
      "		My Second Line",
      "    Third Line",
      "",
      "123",
    ],
  },
  "output": {
    "selectionEnd": 20,
    "selectionStart": 9,
    "value": [
      "My First Line",
      "    My Second Line",
      "    Third Line",
      "",
      "123",
    ],
  },
}
`;

exports[`Editor Commands - UnShiftCommand multiple lines 3`] = `
{
  "config": {
    "insertSpaces": true,
    "tabSize": 4,
  },
  "input": {
    "selectionEnd": 31,
    "selectionStart": 9,
    "value": [
      "My First Line",
      "		My Second Line",
      "    Third Line",
      "",
      "123",
    ],
  },
  "output": {
    "selectionEnd": 33,
    "selectionStart": 9,
    "value": [
      "My First Line",
      "    My Second Line",
      "    Third Line",
      "",
      "123",
    ],
  },
}
`;

exports[`Editor Commands - UnShiftCommand multiple lines 4`] = `
{
  "config": {
    "insertSpaces": true,
    "tabSize": 4,
  },
  "input": {
    "selectionEnd": 15,
    "selectionStart": 3,
    "value": [
      "My First Line",
      "		My Second Line",
      "    Third Line",
      "",
      "123",
    ],
  },
  "output": {
    "selectionEnd": 15,
    "selectionStart": 3,
    "value": [
      "My First Line",
      "    My Second Line",
      "    Third Line",
      "",
      "123",
    ],
  },
}
`;

exports[`Editor Commands - UnShiftCommand multiple lines 5`] = `
{
  "config": {
    "insertSpaces": true,
    "tabSize": 4,
  },
  "input": {
    "selectionEnd": 33,
    "selectionStart": 15,
    "value": [
      "My First Line",
      "		My Second Line",
      "    Third Line",
      "",
      "123",
    ],
  },
  "output": {
    "selectionEnd": 33,
    "selectionStart": 15,
    "value": [
      "My First Line",
      "    My Second Line",
      "Third Line",
      "",
      "123",
    ],
  },
}
`;

exports[`Editor Commands - UnShiftCommand multiple lines 6`] = `
{
  "config": {
    "insertSpaces": true,
    "tabSize": 4,
  },
  "input": {
    "selectionEnd": 50,
    "selectionStart": 0,
    "value": [
      "My First Line",
      "		My Second Line",
      "    Third Line",
      "",
      "123",
    ],
  },
  "output": {
    "selectionEnd": 48,
    "selectionStart": 0,
    "value": [
      "My First Line",
      "    My Second Line",
      "Third Line",
      "",
      "123",
    ],
  },
}
`;
